### PR TITLE
Simplify communication with the content script

### PR DESCRIPTION
**Pull request by psrpinto**
Previously, communication between the content script and the app was done in an way that made it difficult to send commands to the content script, or listen to events coming from the content script.

I originally implemented in a not-great way, and then encountered difficulties, mainly because we would need to copy-paste a bunch of code whenever we wanted to communicate with the content script from a component. This PR addresses these issues, as described below.

I'm doing this now so that I don't need to do the above-mentioned copy-pasting to implement importing the header.

----

There are two types of communication between the app and the content script:

- Commands: when the app tells the content script to do something, e.g. enable the "selection boxes" highlighting, for when the user is selecting elements.
- Events: when something happened in the content script that needs to be communicated to the app, e.g. the user selected an element.

In short:

- Commands: app → content script
- Events: content script → app

## Commands
Sending a command to the content script can now be done as follows:

```ts
import { CommandTypes, sendCommandToContent } from '@/bus/Command';

const response = await sendCommandToContent( {
    type: CommandTypes.NavigateTo,
    payload: {
        url: 'https://example.com'
    },
} );
```

## Events
The content script can send an event to the app as follows:

```ts
import { EventTypes, sendEventToApp } from '@/bus/Event';

void sendEventToApp( {
	type: EventTypes.OnElementClick,
	payload: { content },
} );
```

Events can then be listened to from any component by leveraging the `<ContentEventHandler/>` component. Note this component doesn't render anything, it's just a means to have a callback be called whenever an event is received:

```tsx
import { EventTypes } from '@/bus/Event';
import { ContentEventHandler } from '@/ui/Blueprints/ContentEventHandler;

export function ExampleComponent() {
    return (
        <div>
            <ContentEventHandler
                eventType={ EventTypes.OnElementClick }
                onEvent={ async ( event ) => {
                    const html = event.event.payload.content;
                    // Do something with the event.
                }}
            />
        </div>
    );
}
```

When not in a component, events can be listened to as follows:

```ts
import { startListening } from '@/bus/Bus';
import { CommandTypes } from '@/bus/Command';

startListening( CommandTypes.GetCurrentPageInfo, ( event ) => {
    // Do something with the event.
} );
```

## Sending a response

With both commands and events, a response can be sent to the sender of the command/event:

```ts
import { startListening } from '@/bus/Bus';
import { CommandTypes } from '@/bus/Command';

startListening( CommandTypes.GetCurrentPageInfo, ( event ) => {
	event.sendResponse( {
		url: document.documentURI,
		title: document.title,
	} );
} );
```

### CODE REVIEW

File: src/bus/Bus.ts [hunk 1]
```diff
@@ -0,0 +1,35 @@
+import MessageSender = browser.runtime.MessageSender;
+import { EventType, EventWithNamespace, EventWithResponder } from '@/bus/Event';
+import { CommandType } from '@/bus/Command';
+
+export const Namespace = 'TRY_WORDPRESS';
```
Comment (ashfame) [inline src/bus/Bus.ts:5]: Right now, dependencies of Bus, both Command and Event also import from Bus for getting the Namespace const. So we have a bit of a circular dependency here. Not a big deal, but perhaps we can define it in a separate file and then import it in all 3 files.

**Review by ashfame — APPROVED** [review]
Nice work! AppBus is much cleaner now.

File: src/bus/Bus.ts (hunk 1 above)
Comment (psrpinto) [reply src/bus/Bus.ts:5]: Good point. I will take care of that in another PR, as I'd like to get this merged.

**Comment by psrpinto** [comment]
Thanks for reviewing!

**Pull request merged**